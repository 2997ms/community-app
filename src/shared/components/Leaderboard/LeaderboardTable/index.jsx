/**
 * Leaderboard Table
 *
 * Description:
 *   Table that displays user ranking
 *
 * Usage:
 *   <LeaderboardTable competitors={this.props.competitors} />
 *
 * Props:
 *   - competitors (required): Array of Top Coder user object, with
 *     the following properties:
 *      - rank: Number, required. User current ranking in the leaderboard
 *      - avatarUrl: String. URL for the user's profile picture
 *        This will default to the default user avatar is undefined
 *      - user.handle: String, required. User handle
 *      - wins: Number, required. The number of challenge the user won
 *      - project_result.final_score: Number, required. The user's current score
 */

import config from 'utils/config';
import React from 'react';
import PT from 'prop-types';
import Avatar from 'components/Avatar';
import { defaultAvatarStyles } from '../avatarStyles';

import './styles.scss';

export default function LeaderboardTable(props) {
  const {
    competitors,
  } = props;

  const renderTableRows = comps => (
    comps.map(competitor => (
      <tr key={competitor.rank}>
        <td styleName="col-rank">{competitor.rank}</td>
        <td styleName="col-avatar">
          <span styleName="leaderboard-avatar">
            <Avatar url={competitor.avatarUrl} customStyles={defaultAvatarStyles} />
          </span>
        </td>
        <td styleName="col-handle">
          <a href={`${config.URL.BASE}/members/${competitor['user.handle']}/`}>{competitor['user.handle']}</a>
          <div styleName="winnings-info">
            <span>{competitor['project_result.final_score']} points</span>
            <span>{competitor.wins} challenges</span>
          </div>
        </td>
        <td styleName="col-challenges">{competitor.wins}</td>
        <td styleName="col-points">{competitor['project_result.final_score']}</td>
      </tr>
    ))
  );

  return (
    <table styleName="LeaderboardTable">
      <thead>
        <tr>
          <th styleName="col-rank">Rank</th>
          <th>&nbsp;</th>
          <th styleName="col-handle">Handle</th>
          <th styleName="col-challenges"># of Challenges</th>
          <th styleName="col-points">Points</th>
        </tr>
      </thead>
      <tbody>
        {renderTableRows(competitors)}
      </tbody>
    </table>
  );
}

const CompetitorShape = PT.shape({
  rank: PT.number.isRequired,
  avatarUrl: PT.string,
  'user.handle': PT.string.isRequired,
  wins: PT.number.isRequired,
  'project_result.final_score': PT.number.isRequired,
});

LeaderboardTable.propTypes = {
  competitors: PT.arrayOf(CompetitorShape).isRequired,
};
