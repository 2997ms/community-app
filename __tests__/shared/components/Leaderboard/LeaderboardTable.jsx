import React from 'react';
import Renderer from 'react-test-renderer/shallow';
import LeaderboardTable from 'components/Leaderboard/LeaderboardTable';

const mockUserObj = {
  rank: 1,
  avatarUrl: 'some.fake.string',
  'user.handle': 'overachieving.mofo',
  wins: 99999,
  'project_result.final_score': 9999999999,
};

const mockLeaderboardData = [mockUserObj, mockUserObj, mockUserObj, mockUserObj];

test('Matches shallow shapshot', () => {
  const renderer = new Renderer();
  renderer.render((
    <LeaderboardTable competitors={mockLeaderboardData} />
  ));
  expect(renderer.getRenderOutput()).toMatchSnapshot();
});
