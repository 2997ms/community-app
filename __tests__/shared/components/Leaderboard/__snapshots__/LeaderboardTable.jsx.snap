// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches shallow shapshot 1`] = `
<table
  className="src-shared-components-Leaderboard-LeaderboardTable-___styles__LeaderboardTable___3gXz5"
>
  <thead>
    <tr>
      <th
        className="src-shared-components-Leaderboard-LeaderboardTable-___styles__col-rank___2K_CR"
      >
        Rank
      </th>
      <th>
        Â 
      </th>
      <th
        className="src-shared-components-Leaderboard-LeaderboardTable-___styles__col-handle___3Sr8G"
      >
        Handle
      </th>
      <th
        className="src-shared-components-Leaderboard-LeaderboardTable-___styles__col-challenges___29htm"
      >
        # of Challenges
      </th>
      <th
        className="src-shared-components-Leaderboard-LeaderboardTable-___styles__col-points___10Lis"
      >
        Points
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        className="src-shared-components-Leaderboard-LeaderboardTable-___styles__col-rank___2K_CR"
      >
        1
      </td>
      <td
        className="src-shared-components-Leaderboard-LeaderboardTable-___styles__col-avatar___1xu8j"
      >
        <span
          className="src-shared-components-Leaderboard-LeaderboardTable-___styles__leaderboard-avatar___3Ly7x"
        >
          <ThemedAvatar
            composeTheme="deeply"
            customStyles={null}
            mapThemrProps={[Function]}
            theme={
              Object {
                "avatar": "src-shared-components-Leaderboard-___avatarStyles__default___1bHTz",
              }
            }
            url={null}
          />
        </span>
      </td>
      <td
        className="src-shared-components-Leaderboard-LeaderboardTable-___styles__col-handle___3Sr8G"
      >
        <a
          href="https://www.topcoder-dev.com/members/undefined/"
        />
        <div
          className="src-shared-components-Leaderboard-LeaderboardTable-___styles__winnings-info___1Fiyn"
        >
          <span>
             points
          </span>
          <span>
             challenges
          </span>
        </div>
      </td>
      <td
        className="src-shared-components-Leaderboard-LeaderboardTable-___styles__col-challenges___29htm"
      />
      <td
        className="src-shared-components-Leaderboard-LeaderboardTable-___styles__col-points___10Lis"
      />
    </tr>
    <tr>
      <td
        className="src-shared-components-Leaderboard-LeaderboardTable-___styles__col-rank___2K_CR"
      >
        1
      </td>
      <td
        className="src-shared-components-Leaderboard-LeaderboardTable-___styles__col-avatar___1xu8j"
      >
        <span
          className="src-shared-components-Leaderboard-LeaderboardTable-___styles__leaderboard-avatar___3Ly7x"
        >
          <ThemedAvatar
            composeTheme="deeply"
            customStyles={null}
            mapThemrProps={[Function]}
            theme={
              Object {
                "avatar": "src-shared-components-Leaderboard-___avatarStyles__default___1bHTz",
              }
            }
            url={null}
          />
        </span>
      </td>
      <td
        className="src-shared-components-Leaderboard-LeaderboardTable-___styles__col-handle___3Sr8G"
      >
        <a
          href="https://www.topcoder-dev.com/members/undefined/"
        />
        <div
          className="src-shared-components-Leaderboard-LeaderboardTable-___styles__winnings-info___1Fiyn"
        >
          <span>
             points
          </span>
          <span>
             challenges
          </span>
        </div>
      </td>
      <td
        className="src-shared-components-Leaderboard-LeaderboardTable-___styles__col-challenges___29htm"
      />
      <td
        className="src-shared-components-Leaderboard-LeaderboardTable-___styles__col-points___10Lis"
      />
    </tr>
    <tr>
      <td
        className="src-shared-components-Leaderboard-LeaderboardTable-___styles__col-rank___2K_CR"
      >
        1
      </td>
      <td
        className="src-shared-components-Leaderboard-LeaderboardTable-___styles__col-avatar___1xu8j"
      >
        <span
          className="src-shared-components-Leaderboard-LeaderboardTable-___styles__leaderboard-avatar___3Ly7x"
        >
          <ThemedAvatar
            composeTheme="deeply"
            customStyles={null}
            mapThemrProps={[Function]}
            theme={
              Object {
                "avatar": "src-shared-components-Leaderboard-___avatarStyles__default___1bHTz",
              }
            }
            url={null}
          />
        </span>
      </td>
      <td
        className="src-shared-components-Leaderboard-LeaderboardTable-___styles__col-handle___3Sr8G"
      >
        <a
          href="https://www.topcoder-dev.com/members/undefined/"
        />
        <div
          className="src-shared-components-Leaderboard-LeaderboardTable-___styles__winnings-info___1Fiyn"
        >
          <span>
             points
          </span>
          <span>
             challenges
          </span>
        </div>
      </td>
      <td
        className="src-shared-components-Leaderboard-LeaderboardTable-___styles__col-challenges___29htm"
      />
      <td
        className="src-shared-components-Leaderboard-LeaderboardTable-___styles__col-points___10Lis"
      />
    </tr>
    <tr>
      <td
        className="src-shared-components-Leaderboard-LeaderboardTable-___styles__col-rank___2K_CR"
      >
        1
      </td>
      <td
        className="src-shared-components-Leaderboard-LeaderboardTable-___styles__col-avatar___1xu8j"
      >
        <span
          className="src-shared-components-Leaderboard-LeaderboardTable-___styles__leaderboard-avatar___3Ly7x"
        >
          <ThemedAvatar
            composeTheme="deeply"
            customStyles={null}
            mapThemrProps={[Function]}
            theme={
              Object {
                "avatar": "src-shared-components-Leaderboard-___avatarStyles__default___1bHTz",
              }
            }
            url={null}
          />
        </span>
      </td>
      <td
        className="src-shared-components-Leaderboard-LeaderboardTable-___styles__col-handle___3Sr8G"
      >
        <a
          href="https://www.topcoder-dev.com/members/undefined/"
        />
        <div
          className="src-shared-components-Leaderboard-LeaderboardTable-___styles__winnings-info___1Fiyn"
        >
          <span>
             points
          </span>
          <span>
             challenges
          </span>
        </div>
      </td>
      <td
        className="src-shared-components-Leaderboard-LeaderboardTable-___styles__col-challenges___29htm"
      />
      <td
        className="src-shared-components-Leaderboard-LeaderboardTable-___styles__col-points___10Lis"
      />
    </tr>
  </tbody>
</table>
`;
